### DISCUSSION TITLE
Try example with websocket

### DISCUSSION THREAD

**Discussion by Aliw7979** [juanmc2005/diart#123] [Q&A] [answered]
Hi, I need to know how to send bytes of data to WebSocketAudioSource().

Created websocket server like this:
```

from diart import OnlineSpeakerDiarization
from diart.sources import WebSocketAudioSource,MicrophoneAudioSource
from diart.inference import RealTimeInference

pipeline = OnlineSpeakerDiarization()
source = WebSocketAudioSource(pipeline.config.sample_rate, "localhost", 7007)
inference = RealTimeInference(pipeline, source, do_plot=True)
inference.attach_hooks(lambda ann_wav: source.send(ann_wav[0].to_rttm()))
prediction = inference()
```

And send data bytes like this:

```
import asyncio
import websockets
import sys
import wave
import base64
import time

async def run_test(uri):
    async with websockets.connect(uri) as websocket:

        wf = wave.open(sys.argv[1], "rb")
        while True:
            encoded_string = base64.encodebytes(wf.readframes(1024))
            print(encoded_string)
            if len(encoded_string) == 0:
                 break
            
            await websocket.send(str(encoded_string))
            time.sleep(0.1)

asyncio.run(run_test('ws://localhost:7007'))

```
but get this error: 

ValueError: buffer size must be a multiple of element size

**Comment by juanmc2005** [comment, marked as answer]
Hi @Aliw7979, could you share the entire stacktrace so I can know where the error is coming from?

**Reply by Aliw7979** [reply]
Sure,
```

Streaming localhost:7007: 0chunk [00:00, ?chunk/s]connection handler failed
Traceback (most recent call last):
  File "/home/ali/miniconda3/envs/diart/lib/python3.8/site-packages/websockets/legacy/server.py", line 236, in handler
    await self.ws_handler(self)
  File "/home/ali/miniconda3/envs/diart/lib/python3.8/site-packages/diart/sources.py", line 186, in _ws_handler
    samples= np.frombuffer(byte_samples, dtype=np.float64)
ValueError: buffer size must be a multiple of element size
```

I send 16k wav file,

**Reply by juanmc2005** [reply]
I think there may be a problem with your encoding.
I recently re-wrote the `WebSocketAudioSource` and added encoding and decoding functions for convenience: `diart.utils.encode_audio` and `diart.utils.decode_audio`.

Could you install from branch `fix/ws` and try again using these functions?

**Reply by Aliw7979** [reply]
First of all I appreciate your responsiveness, and I updated diart with `fix/ws` and my problem was almost fixed.
Result:

![Screenshot from 2023-03-14 16-28-01](https://user-images.githubusercontent.com/60917259/225008968-3341e2fc-67a6-4ca2-9c35-ed5dfef251a4.png)

But I have to ask that does received messages correspond to the plot?
And what are those exceptions ? I think those exceptions occurred because of not receiving continuous bytes from client.

**Reply by Aliw7979** [reply]
By the way I have to mention that plot is showing correct result and speaker detected accurately.

**Reply by juanmc2005** [reply]
Glad to hear it's working!

> But I have to ask that does received messages correspond to the plot?

Yes they are the same ones shown in the plot, the difference lies in the fact that the plot merges short segments with pauses shorter than 50ms. This over-segmentation occurs because the system outputs contiguous 500ms predictions.

> And what are those exceptions ? I think those exceptions occurred because of not receiving continuous bytes from client.

I had similar issues when I implemented the client with `websockets`. As soon as I tried with `websocket-client` it started working without a problem. I suspect it has something to do with handshakes.
Could you try using `diart.client` and see if you still get these errors ?

Another important thing: make sure to disable plotting on the server side.
I got errors when plotting server-side because both the server and the plot needed to run on the main thread.

**Reply by Aliw7979** [reply]
I will try it with `diart.client`. Is it available in `develop`  branch? Because I couldn't find `diart.client` in `fix/ws` branch.

**Reply by juanmc2005** [reply]
Hi @Aliw7979, yes this is now available on the `develop` branch and will be part of the soon-to-be-realeased v0.7

**Comment by Aliw7979** [comment]
Hi @juanmc2005, I couldn't use diart.client through terminal:
results:
 
![Screenshot from 2023-04-16 17-16-28](https://user-images.githubusercontent.com/60917259/232315926-31a1d40b-4eee-4fb5-959c-ebbed0ea4d36.png)

by the way, doesn't diart support multiple client?

**Reply by juanmc2005** [reply]
Hi @Aliw7979, it seems that there's a problem with the sample rate, which is 16000 by default. Could you verify the supported sample rate of your microphone? You can also select other devices with `diart.stream microphone:<ID>`.

You can change the sample rate of the client with `diart.client ... --sample-rate 32000`.

Don't worry about conflicting sample rates down the line, diart will attempt to resample the audio stream to match the requirements of `pyannote/segmentation` if they differ. I think you even get a warning when it does.

> by the way, doesn't diart support multiple client?

This is not yet supported using a single server, but you could spawn multiple instances of diart.serve in different ports. Just remember that to reuse a port for a new conversation you have to restart the server (or reset the pipeline).

The main problem with supporting multiple clients is that you need to make sure that each client is processed with its own diarization pipeline (if 1 client = 1 conversation) or that you merge the audio from different clients (if many clients = 1 conversation).
The latter is also a bit more tricky because you'd need to align the timestamps and sample rates from different microphones.

If you're interested in contributing, it would be awesome to add this feature. I can also help you accelerate development and get a PR merged.

**Reply by Aliw7979** [reply]
I will try diart.client again and send updates.

> If you're interested in contributing, it would be awesome to add this feature. I can also help you accelerate development and get a PR merged.

Actually, I had a plan to add this feature besides an ASR to diart, it is important to me either. As we have the same goal, I'm eager to contribute.

**Reply by juanmc2005** [reply]
That's amazing! Don't hesitate to create a draft PR so we can discuss the more technical bits.

On the ASR side, I already started working on this with Whisper. I actually wrote a [blog post](https://medium.com/better-programming/color-your-captions-streamlining-live-transcriptions-with-diart-and-openais-whisper-6203350234ef) on how to combine it with diart that could be a good starting point.
There are still many things to improve, like using a sliding window with a certain overlap and merging contiguous transcriptions.

**Reply by Aliw7979** [reply]
Ok then, can you give me permission to create draft pull request?

> On the ASR side, I already started working on this with Whisper. I actually wrote a [blog post](https://medium.com/better-programming/color-your-captions-streamlining-live-transcriptions-with-diart-and-openais-whisper-6203350234ef) on how to combine it with diart that could be a good starting point.

Thank you for more information, I am going to work on WebSocket first then switch on ASR side.

**Reply by juanmc2005** [reply]
I just opened PR #144 for the Whisper-based ASR feature, although it's still "speaker-blind".
You can go ahead and open a normal pull request and I'll convert it to draft once it's there.
